### ISSUE TITLE
s3-put fails with error 403

### ISSUE THREAD

**Issue by TheCallSign** [milliHQ/terraform-aws-next-js#40]
The s3-put bash script seems to fail to upload the static files:

```
Error: Error running command './s3-put -r us-east-1 -T /home/thecallsign/code/.next-tf/static-website-files.zip /next-tf-deploy-source20210118142043963400000001/static-website-files.zip': exit status 22. 
Output:   
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0 1618k    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 403 Forbidden
```

All the env vars are set correctly, and after a couple hours trying to figure out what was causing the issue I forked this repo to bypass s3-put completely and just use a terraform config block. [Source](https://github.com/TheCallSign/terraform-aws-next-js/blob/b1c94628e0dbb864e48335916752ffe9a0f02b54/modules/statics-deploy/main.tf#L204)

Things seems to be working. Is there a reason to be using the bash script instead?

**Comment by ofhouse** [comment]
Ah the good old bash workaround 🙈
From a Terraform perspective the `static-website-files.zip` file is not a resource that Terraform manages (Because it gets deleted by the static deployment Lamda after upload).
Because the file gets deleted Terraform performs an upload on every apply (Instead of matching the etag) which can cause an unintended deployment.
So we outsourced the upload into a seperate script.

Originally this was part of the `tf-next` package but because our Terraform CI does not support Node.js we switched to this Bash version for doing the job.

The Bash script has dependencies to `OpenSSL` and `cUrl` so they must be installed for running the command:
```sh
openssl version
> LibreSSL 2.8.3

curl --version
> curl 7.64.1
```

Easiest solution would be to persist the `static-website-files.zip` file on S3 but I decided against it because it can be pretty large for big sites (with images etc.) and would cost you money while not offering any benefit.

Another solution would be an option to switch to the [AWS CLI](https://aws.amazon.com/cli/) for doing the task.

**Comment by TheCallSign** [comment]
Perhaps leaving a small fixed sized tombstone with the same etag in the object's place after the lambda unpacks the zip might work?


Here are my versions

```
$ uname -a
Linux signz 5.9.0-4-amd64 #1 SMP Debian 5.9.11-1 (2020-11-27) x86_64 GNU/Linux
$ openssl version
OpenSSL 1.1.1h  22 Sep 2020
$ curl --version
curl 7.72.0 (x86_64-pc-linux-gnu) libcurl/7.72.0 OpenSSL/1.1.1h zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.8.0 nghttp2/1.42.0 librtmp/2.3
```

**Comment by TheCallSign** [comment]
Would using a terraform resource like
```
resource "aws_s3_bucket_object" "static_s3_upload" {
  bucket = aws_s3_bucket.static_upload.id
  key = basename(var.static_files_archive)
  source = abspath(var.static_files_archive)
  etag   = filemd5(var.static_files_archive)

  lifecycle {  
    ignore_changes = ["etag"] # ignore changes to the contents of the object
  }
  # Make sure this only runs when the bucket and the lambda trigger are setup
  depends_on = [
    aws_s3_bucket_notification.on_create
  ]
}
```

and replacing the object with an empty object in its place work?

**Comment by TheCallSign** [comment]
Nevermind - that probably defeats the purpose of the etag attribute as a whole. My bad.

**Comment by ofhouse** [comment]
I like the idea in general, but ETag is always calculated by S3 itself based on the content and is not modifiable by from the API/SDK.
So sending a PUT operation on the object with an empty body would always result in a new object with a different ETag.

So the modifier key must be placed in the object meta data and cannot be ETag.
But I don't know if Terraform would check for changes in the meta data of the object or if the resource would only make a comparison of the ETag.

**Comment by ofhouse** [comment]
The thing is that Terraform is possibly the wrong platform for doing deployments like this.
That is also where I see this module in the future: Instead of running apply on every deployment, Terraform should only create a Platform where a tool like `tf-next` can then push the deployments to.

Internally the deployment should then be handled by [CodeDeploy](https://aws.amazon.com/codedeploy/), [Waypoint](https://www.hashicorp.com/blog/announcing-waypoint) or an internal service (similar how the static upload Lambda works today).

**Comment by chamilad** [comment]
@ofhouse  I'm coming across this issue on a MacOS with the exact versions (and implementations) you mentioned above. Using the AWS CLI works exactly as you mentioned. Thanks for that!

Would you be open to using a flag to use either of these approaches? I opened PR #83 with my changes. WDYT?

**Comment by ofhouse** [comment]
Yep, AWS CLI as fallback is currently the best workaround, thanks for adding this as option 👍 

I will keep this issue open because I think the task should eventually handled by a native approach rather than external tooling (Bash/AWS CLI).
Need to reevaluate this further after the `0.9.0` release.

**Comment by rodrigogazzaneo** [comment]
Hello @ofhouse ,

I have the same problem uploading the static-website-files.zip.

I tried to do with AWSCLI like that:

```

> module "tf_next" {
>   source = "dealmore/next-js/aws"
>   use_awscli_for_static_upload = true
>   deployment_name = "terraform-next-js-example-complete"
> 
>   providers = {
>     aws.global_region = aws.global_region
>   }

```
But still getting an error:

> Error: local-exec provisioner error
> │
> │   with module.tf_next.module.statics_deploy.null_resource.static_s3_upload_awscli[0],
> │   on .terraform/modules/tf_next/modules/statics-deploy/main.tf line 242, in resource "null_resource" "static_s3_upload_awscli":
> │  242:   provisioner "local-exec" {
> │
> │ Error running command 'aws s3 cp --region us-east-1 /home/my-app/.next-tf/static-website-files.zip
> │ s3://next-tf-deploy-source20210622211329320000000003/static-website-files.zip': exit status 1. Output: upload failed: .next-tf/static-website-files.zip to
> │ s3://next-tf-deploy-source20210622211329320000000003/static-website-files.zip Unable to locate credentials

LibreSSL and Curl are like this:

LibreSSL 3.3.3
curl 7.68.0

Am I missing anything?

Thank you,
Rodrigo

**Comment by ofhouse** [comment]
Hi @rodrigogazzaneo,
the error message at the end says `Unable to locate credentials`.

So I guess that the Access key ID and Secret Access key are not exposed in the Terminal session.
Easiest way is do do this by setting it via the environment variables:
```
export AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
export AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

More options to provide the credentials are described in the AWS CLI documentation: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html

**Comment by rodrigogazzaneo** [comment]
Hi,
Thanks for your answer.

I found out the AWS credentials were set to the user (set by `aws configure` - with no sudo) and not to the root. When I was trying to do `sudo terraform apply`, there were no credentials.

So I just made a `sudo aws configure` and set credentials to the root.

I think exposing it via terminal session would work with 
```
sudo export AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
sudo export AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```
Thank you for your help.
Best regards,
Rodrigo
